src/SAWScript/Proof.hs

```diff
@@ -923,25 +1549,37 @@ setProofTimeout to ps = ps { _psTimeout = Just to }
 startProof :: ProofGoal -> Pos -> Maybe ProgramLoc -> Text -> IO ProofState
 startProof g pos ploc rsn =
   do start <- getCurrentTime
-     pure (ProofState [g] (goalProp g,pos,ploc,rsn) mempty Nothing passthroughEvidence start)
+     pure (ProofState [g] (goalSequent g,pos,ploc,rsn) mempty Nothing passthroughEvidence start)
 
 -- | Attempt to complete a proof by checking that all subgoals have been discharged,
 --   and validate the computed evidence to ensure that it supports the original
 --   proposition.  If successful, return the completed @Theorem@ and a summary
 --   of solver resources used in the proof.
-finishProof :: SharedContext -> TheoremDB -> ProofState -> IO ProofResult
-finishProof sc db ps@(ProofState gs (concl,loc,ploc,rsn) stats _ checkEv start) =
+--
+--   If the final boolean argument is False, the resulting theorem will not be
+--   recored in the theorem database. This should only be done when you are
+--   sure that the theorem will not be used as part of the proof of other theorems,
+--   or later steps will fail.
+finishProof ::
+  SharedContext ->
+  TheoremDB ->
+  Prop ->
+  ProofState ->
+  Bool {- ^ should we record the theorem in the database? -} ->
+  IO ProofResult
+finishProof sc db conclProp ps@(ProofState gs (concl,loc,ploc,rsn) stats _ checkEv start) recordThm =
   case gs of
     [] ->
       do e <- checkEv []
-         (deps,sy) <- checkEvidence sc db e concl
+         let e' = NormalizeSequentEvidence concl e
```

Comment:
This line seems to be the culprit for the proof slowdowns.  It forces a sequent normalization step, which can apparently be pretty expensive in some cases. The step is only sometimes necessary, so I guess I should figure out how to apply it just when needed.